ext/semian_cb_data/semian_cb_data.c

```diff
@@ -0,0 +1,900 @@
+
+  if (-1 == semtimedop(ptr->semid,&increment,1 , &ts)) {
+    rb_raise(eInternal, "error with semop unlocking, errno: %d (%s)", errno, strerror(errno));
+    retval=Qfalse;
+  } else
+    retval=Qtrue;
+
+  ptr->lock_triggered = false;
+  //rb_warn("semop unlock success");
+  return (void *)retval;
+}
+
+
+
+
+/*
+  Acquire memory by getting shmid, and then attaching it to a memory location,
+    requires semaphore for locking/unlocking to be setup
+*/
+static VALUE
+semian_cb_data_acquire_memory(VALUE self, VALUE permissions)
+{
+  semian_cb_data *ptr;
+  TypedData_Get_Struct(self, semian_cb_data, &semian_cb_data_type, ptr);
+
+  if (-1 == ptr->semid){
+    rb_raise(eInternal, "semid not set, errno %d: (%s)", errno, strerror(errno));
+    return self;
+  }
+  if (TYPE(permissions) != T_FIXNUM)
+    rb_raise(rb_eTypeError, "expected integer for permissions");
+
+  if (!semian_cb_data_lock(self))
+    return self;
+
+  key_t key = ptr->key;
+  if (-1 == (ptr->shmid = shmget( key,
+                                  2*sizeof(int) + ptr->arr_max_size * sizeof(double),
+                                  IPC_CREAT | IPC_EXCL | FIX2LONG(permissions)))) {
+    if (errno == EEXIST)
+      ptr->shmid = shmget(key, ptr->arr_max_size, IPC_CREAT);
+  }
+  if (-1 == ptr->shmid) {
+    rb_raise(eSyscall, "shmget() failed to acquire a memory shmid with key %d, size %zu, errno %d (%s)", key, ptr->arr_max_size, errno, strerror(errno));
+  } else {
+    //rb_warn("shmget: successfully got memory id with key %d, shmid %d, size %zu", key, ptr->shmid, ptr->arr_max_size);
+  }
+
+  if (0 == ptr->shm_address) {
+    ptr->shm_address = shmat(ptr->shmid, (void *)0, 0);
+    if (((void*)-1) == ptr->shm_address) {
+      rb_raise(eSyscall, "shmat() failed to attach memory with shmid %d, size %zu, errno %d (%s)", ptr->shmid, ptr->arr_max_size, errno, strerror(errno));
+      ptr->shm_address = 0;
+    } else {
+      //rb_warn("shmat: successfully attached shmid %d to %p", ptr->shmid, ptr->shm_address);
+      shared_cb_data *data = ptr->shm_address;
+      data->successes = 0;
+      data->arr_length = 0;
+      int i=0;
+      for (; i< data->arr_length; ++i)
+        data->errors[i]=0;
+
+    }
+  }
+
+  semian_cb_data_unlock(self);
+  return self;
+}
+
+static void
+semian_cb_data_delete_memory_inner (semian_cb_data *ptr)
+{
+  if (0 != ptr->shm_address){
+    if (-1 == shmdt(ptr->shm_address)) {
+      rb_raise(eSyscall,"shmdt: no attached memory at %p, errno %d (%s)", ptr->shm_address, errno, strerror(errno));
+    } else {
+      rb_warn("shmdt: successfully detached memory at %p", ptr->shm_address);
+    }
+    ptr->shm_address = 0;
+  }
+
+  if (-1 != ptr->shmid) {
+    // Once IPC_RMID is set, no new attaches can be made
+    if (-1 == shmctl(ptr->shmid, IPC_RMID, 0)) {
+      if (errno == EINVAL) {
+        ptr->shmid = -1;
+      } {
+        rb_raise(eSyscall,"shmctl: error removing memory with shmid %d, errno %d (%s)", ptr->shmid, errno, strerror(errno));
```

Comment:
It doesn't quite mark the memory, it starts an async action. It's more of a marking for deletion.